pvlib/tools.py

```diff
@@ -425,3 +426,58 @@ def _golden_sect_DataFrame(params, VL, VH, func):
             raise Exception("EXCEPTION:iterations exceeded maximum (50)")
 
     return func(df, 'V1'), df['V1']
+
+
+def enforce_numpy_arrays(f):
+    """ This decorator function will convert any inputted pandas dataframe or
+    series to a numpy array, and make sure to convert outtputed numpy arrays
+    to pandas series if any input was a pandas structure """
+
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        # Convert all inputs to numpy arrays if pandas series or dataframe
+        is_return_type_pandas = False
+
+        new_args = []
+        for arg in args:
+            if isinstance(arg, pd.Series):
+                is_return_type_pandas = True
+                new_arg = arg.values
+                new_args.append(new_arg)
+            elif isinstance(arg, pd.DataFrame):
+                is_return_type_pandas = True
+                new_arg = arg.values.ravel()  # make sure that 1D
+                new_args.append(new_arg)
+            else:
+                new_args.append(arg)
+        new_args = tuple(new_args)
+
+        new_kwargs = {}
+        for key, val in kwargs.items():
+            if isinstance(val, pd.DataFrame) or isinstance(val, pd.Series):
+                is_return_type_pandas = True
+                new_val = val.values
+                new_kwargs[key] = new_val
+            else:
+                new_kwargs[key] = val
+
+        # Run function
+        outputs = f(*new_args, **new_kwargs)
+
+        # If inputs were provided as pandas series or dataframe, convert
+        # outputted numpy arrays to pandas series
+        new_outputs = []
+        if is_return_type_pandas:
+            for output in outputs:
+                if isinstance(output, np.ndarray):
+                    new_output = pd.Series(output)
```

Comment:
Looks like this function ignores the input index. `Series (with meaningful index) --> 1D array --> Series (without meaningful index)` is not very friendly. Arguably better to just return the 1D array and avoid the complication of a decorator.

Reply:
Thanks @wholmgren . In the specific case of the ``pvfactors_timeseries()`` function, the ``timestamps`` index is required as a separate argument, but your point stays valid in any case.
Would you be ok with the decorating function recording and passing any ``meaningful index`` to potentially outputted Series to answer this concern?

Reply:
I suggest we make a less-general function that's only designed to handle the inputs for `pvfactors_timeseries`: series/1d array in, numpy out, exception otherwise. Then we can step back and assess how we can make it more general for other functions in another issue/pr.

Reply:
Thanks @wholmgren . I removed the decorator function and am now doing the checks inside the pvfactors function. Let me know if the new implementation is ok